File: add-two-numbers/add-two-numbers.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    ptr = ListNode.new()
    ans = ptr
    carry = 0
    while l1 && l2 do
        ptr.next = ListNode.new((l1.val + l2.val + carry)%10)
        carry = (l1.val + l2.val + carry)/10
        l1 = l1.next
        l2 = l2.next
        ptr = ptr.next
    end

    while l1 do
        ptr.next = ListNode.new((l1.val + carry)%10)
        carry = (l1.val + carry)/10
        l1 = l1.next
        ptr = ptr.next
    end

    while l2 do
        ptr.next = ListNode.new((l2.val + carry)%10)
        carry = (l2.val + carry)/10
        l2 = l2.next
        ptr = ptr.next
    end

    if carry != 0
        ptr.next = ListNode.new(carry)
    end

    ptr = ans.next

    return ptr
end

File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode* head = new ListNode();
        ListNode* ptr = head;
        while(l1 && l2){
            int val = (l1->val + l2->val+carry)%10;
            carry = (l1->val + l2->val+carry)/10;
            ptr->next =  new ListNode(val);
            l1 = l1->next;
            l2 = l2->next;
            ptr = ptr->next;
        }
        while(l1){
            int val = (l1->val + carry)%10;
            carry = (l1->val + carry)/10;
            ptr->next =  new ListNode(val);
            l1 = l1->next;
            ptr = ptr->next;
        }
        while(l2){
            int val = (l2->val + carry)%10;
            carry = (l2->val + carry)/10;
            ptr->next =  new ListNode(val);
            l2 = l2->next;
            ptr = ptr->next;
        }
        if(carry){
            ptr->next =  new ListNode(carry);
        }

        return head->next;
    }
};